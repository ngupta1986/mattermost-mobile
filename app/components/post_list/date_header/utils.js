// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {DATE_LINE} from 'app/selectors/post_list';

export const isDateLine = (dateString) => dateString.indexOf(DATE_LINE) === 0;
