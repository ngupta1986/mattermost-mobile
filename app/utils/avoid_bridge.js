
import {Platform} from 'react-native';

export default async function avoidNativeBridge(runOptimized, optimized, fallback) {
    if (Platform.OS === 'android' && runOptimized()) { // Only required for Android
        return optimized();
    }

    return await fallback();
}
