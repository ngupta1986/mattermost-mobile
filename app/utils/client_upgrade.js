// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import semver from 'semver';

import {UpgradeTypes} from 'app/constants/view';

import LocalConfig from 'assets/config';

export default function checkUpgradeType(currentVersion, minVersion, latestVersion, logError) {
    let upgradeType = UpgradeTypes.NO_UPGRADE;

    try {
        if (minVersion && semver.lt(currentVersion, minVersion)) {
            upgradeType = UpgradeTypes.MUST_UPGRADE;
        } else if (latestVersion && semver.lt(currentVersion, latestVersion)) {
            if (LocalConfig.EnableForceMobileClientUpgrade) {
                upgradeType = UpgradeTypes.MUST_UPGRADE;
            } else {
                upgradeType = UpgradeTypes.CAN_UPGRADE;
            }
        }
    } catch (error) {
        logError(error.message);
    }

    return upgradeType;
}
