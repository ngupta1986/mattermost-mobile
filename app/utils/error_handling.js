// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {
    Alert,
} from 'react-native';

import {
    setJSExceptionHandler,
    setNativeExceptionHandler,
} from 'react-native-exception-handler';

import {Client4} from 'mattermost-redux/client';
import {logError} from 'mattermost-redux/actions/errors';
import {close as closeWebSocket} from 'mattermost-redux/actions/websocket';

import {purgeOfflineStore} from 'app/actions/views/root';
import {
    captureException,
    initializeSentry,
    LOGGER_JAVASCRIPT,
    LOGGER_NATIVE,
} from 'app/utils/sentry';

import {app, store} from 'app/mattermost';

const errorHandler = (e, isFatal) => {
    console.warn('Handling Javascript error ', e); // eslint-disable-line no-console
    const {dispatch} = store;

    captureException(e, LOGGER_JAVASCRIPT, store);

    const translations = app.getTranslations();
    dispatch(closeWebSocket());

    if (Client4.getUrl()) {
        dispatch(logError(e));
    }

    if (isFatal) {
        Alert.alert(
            translations['mobile.error_handler.title'],
            translations['mobile.error_handler.description'],
            [{
                text: translations['mobile.error_handler.button'],
                onPress: () => {
                    // purge the store
                    dispatch(purgeOfflineStore());
                },
            }],
            {cancelable: false}
        );
    }
};

const nativeErrorHandler = (e) => {
    console.warn('Handling native error ' + JSON.stringify(e)); // eslint-disable-line no-console
    captureException(e, LOGGER_NATIVE, store);
};

export function initializeErrorHandling() {
    initializeSentry();
    setJSExceptionHandler(errorHandler, false);
    setNativeExceptionHandler(nativeErrorHandler, false);
}
