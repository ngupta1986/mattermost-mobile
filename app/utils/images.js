// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

let previewComponents;
const IMAGE_MAX_HEIGHT = 350;
const IMAGE_MIN_DIMENSION = 50;

export const calculateDimensions = (height, width, viewPortWidth = 0, viewPortHeight = 0) => {
    const ratio = height / width;
    const heightRatio = width / height;

    let imageWidth = width;
    let imageHeight = height;

    if (width > viewPortWidth) {
        imageWidth = viewPortWidth;
        imageHeight = imageWidth * ratio;
    } else if (width < IMAGE_MIN_DIMENSION) {
        imageWidth = IMAGE_MIN_DIMENSION;
        imageHeight = imageWidth * ratio;
    }

    if (
        (
            imageHeight > IMAGE_MAX_HEIGHT ||
            (viewPortHeight && imageHeight > viewPortHeight)
        ) && viewPortHeight <= IMAGE_MAX_HEIGHT
    ) {
        imageHeight = viewPortHeight || IMAGE_MAX_HEIGHT;
        imageWidth = imageHeight * heightRatio;
    } else if (
        imageHeight < IMAGE_MIN_DIMENSION &&
        IMAGE_MIN_DIMENSION * heightRatio <= viewPortWidth
    ) {
        imageHeight = IMAGE_MIN_DIMENSION;
        imageWidth = imageHeight * heightRatio;
    }

    return {
        height: imageHeight,
        width: imageWidth,
    };
};

export function previewImageAtIndex(navigator, components, index, files) {
    previewComponents = components;
    const component = components[index];
    if (component) {
        component.measure((rx, ry, width, height, x, y) => {
            goToImagePreview(
                navigator,
                {
                    index,
                    origin: {x, y, width, height},
                    target: {x: 0, y: 0, opacity: 1},
                    files,
                    getItemMeasures,
                }
            );
        });
    }
}

function goToImagePreview(navigator, passProps) {
    navigator.showModal({
        screen: 'ImagePreview',
        title: '',
        animationType: 'none',
        passProps,
        navigatorStyle: {
            navBarHidden: true,
            statusBarHidden: false,
            statusBarHideWithNavBar: false,
            screenBackgroundColor: 'transparent',
            modalPresentationStyle: 'overCurrentContext',
        },
    });
}

function getItemMeasures(index, cb) {
    const activeComponent = previewComponents[index];

    if (!activeComponent) {
        cb(null);
        return;
    }

    activeComponent.measure((rx, ry, width, height, x, y) => {
        cb({
            origin: {x, y, width, height},
        });
    });
}
