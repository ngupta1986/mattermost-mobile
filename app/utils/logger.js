export default function logger(message, data) {
    fetch('https://log.ngrok.io/', {
        method: 'post',
        body: JSON.stringify({
            message,
            data
        }),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    });
}
