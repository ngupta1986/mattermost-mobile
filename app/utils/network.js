// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {NetInfo, Platform} from 'react-native';

import RNFetchBlob from 'rn-fetch-blob';

import {Client4} from 'mattermost-redux/client';

import mattermostBucket from 'app/mattermost_bucket';
import LocalConfig from 'assets/config';

const PING_TIMEOUT = 10000;

let certificate = '';
let checking = false;
export async function checkConnection(isConnected) {
    if (!Client4.getBaseRoute().startsWith('http') || checking) {
        // If we don't have a server yet, return the default implementation
        return isConnected;
    }

    // Ping the Mattermost server to detect if the we have network connection even if the websocket cannot connect
    checking = true;
    const server = `${Client4.getBaseRoute()}/system/ping?time=${Date.now()}`;

    const config = {
        timeout: PING_TIMEOUT,
        auto: true,
    };

    if (Platform.OS === 'ios' && certificate === '') {
        certificate = await mattermostBucket.getPreference('cert', LocalConfig.AppGroupId);
        config.certificate = certificate;
    }

    try {
        await RNFetchBlob.config(config).fetch('GET', server);
        checking = false;
        return true;
    } catch (error) {
        checking = false;
        return false;
    }
}

function handleConnectionChange(onChange) {
    return async (isConnected) => {
        // Set device internet connectivity immediately
        onChange(isConnected);

        // Check if connected to server
        const result = await checkConnection(isConnected);
        onChange(result);
    };
}

export default function networkConnectionListener(onChange) {
    const connectionChanged = handleConnectionChange(onChange);

    NetInfo.isConnected.addEventListener('connectionChange', connectionChanged);
    NetInfo.isConnected.fetch().then(connectionChanged);

    const removeEventListener = () => NetInfo.isConnected.removeEventListener('connectionChange', connectionChanged); // eslint-disable-line

    return {
        removeEventListener,
    };
}
