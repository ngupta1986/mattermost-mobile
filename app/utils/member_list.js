// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {t} from 'app/utils/i18n';

export const loadingText = {
    id: t('mobile.loading_members'),
    defaultMessage: 'Loading Members...',
};

export function createMembersSections(data) {
    const sections = {};
    data.forEach((d) => {
        const name = d.username;
        const sectionKey = name.substring(0, 1).toUpperCase();

        if (!sections[sectionKey]) {
            sections[sectionKey] = [];
        }

        sections[sectionKey].push(d);
    });

    return sections;
}

export function markSelectedProfiles(profiles, selectedProfiles) {
    return profiles.map((p) => {
        const profile = {...p};
        profile.selected = selectedProfiles[p.id];
        return profile;
    });
}
