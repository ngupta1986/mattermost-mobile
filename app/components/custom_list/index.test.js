// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import Preferences from 'mattermost-redux/constants/preferences';

import {createMembersSections, loadingText} from 'app/utils/member_list';

import CustomList from './index';

describe('CustomList', () => {
    function emptyFunc() {} // eslint-disable-line no-empty-function

    const baseProps = {
        data: [{username: 'username_1'}, {username: 'username_2'}],
        theme: Preferences.THEMES.default,
        searching: false,
        onListEndReached: emptyFunc,
        onListEndReachedThreshold: 0,
        loading: false,
        loadingText,
        listPageSize: 0,
        listInitialSize: 0,
        listScrollRenderAheadDistance: 0,
        showSections: true,
        onRowPress: emptyFunc,
        selectable: true,
        onRowSelect: emptyFunc,
        renderRow: emptyFunc,
        createSections: createMembersSections,
        showNoResults: false,
    };

    test('should match snapshot', () => {
        const wrapper = shallow(
            <CustomList {...baseProps}/>
        );
        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper.state('data')).toEqual({U: baseProps.data});

        // on componentWillReceiveProps
        const newDataProps = [{username: 'username_1'}, {username: 'username_2'}, {username: 'username_3'}];
        wrapper.setProps({data: newDataProps});
        expect(wrapper.state('data')).toEqual({U: newDataProps});
    });

    test('should match return value on buildDataSource', () => {
        const props = {...baseProps};
        const wrapper = shallow(
            <CustomList {...props}/>
        );

        expect(wrapper.instance().buildDataSource(props).data).toEqual({U: [{username: 'username_1'}, {username: 'username_2'}]});
        const newDataProps = [{username: 'aaa_1'}, {username: 'aaa_2'}, {username: 'username_1'}, {username: 'username_2'}, {username: 'username_3'}];
        props.data = newDataProps;
        expect(wrapper.instance().buildDataSource(props).data).toEqual({
            A: [{username: 'aaa_1'}, {username: 'aaa_2'}],
            U: [{username: 'username_1'}, {username: 'username_2'}, {username: 'username_3'}],
        });
    });

    test('should match snapshot, renderSectionHeader', () => {
        const wrapper = shallow(
            <CustomList {...baseProps}/>
        );
        expect(wrapper.instance().renderSectionHeader([], 'section_id')).toMatchSnapshot();
    });

    test('should call props.renderRow on renderRow', () => {
        const props = {...baseProps, renderRow: jest.fn()};
        const wrapper = shallow(
            <CustomList {...props}/>
        );
        wrapper.instance().renderRow({id: 'item_id', selected: true}, 'section_id', 'row_id');
        expect(props.renderRow).toHaveBeenCalledTimes(1);
    });

    test('should match snapshot, renderSeparator', () => {
        const wrapper = shallow(
            <CustomList {...baseProps}/>
        );
        expect(wrapper.instance().renderSeparator('section_id', 'row_id')).toMatchSnapshot();
    });

    test('should match snapshot, renderFooter', () => {
        const props = {...baseProps};
        const wrapper = shallow(
            <CustomList {...props}/>
        );

        // should return null
        expect(wrapper.instance().renderFooter()).toMatchSnapshot();

        // should return element
        wrapper.setProps({loading: true});
        expect(wrapper.instance().renderFooter()).toMatchSnapshot();
    });
});
