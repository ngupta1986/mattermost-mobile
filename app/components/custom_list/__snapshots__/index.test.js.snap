// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomList should match snapshot 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ListViewMock
    dataSource={
      ListViewDataSource {
        "items": 2,
      }
    }
    enableEmptySections={true}
    initialListSize={0}
    onEndReached={[Function]}
    onEndReachedThreshold={0}
    pageSize={0}
    renderFooter={[Function]}
    renderRow={[Function]}
    renderScrollComponent={[Function]}
    renderSectionHeader={[Function]}
    renderSeparator={[Function]}
    scrollRenderAheadDistance={0}
    style={
      Object {
        "backgroundColor": "#ffffff",
        "flex": 1,
      }
    }
  />
</Component>
`;

exports[`CustomList should match snapshot, renderFooter 1`] = `null`;

exports[`CustomList should match snapshot, renderFooter 2`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "height": 70,
      "justifyContent": "center",
    }
  }
>
  <FormattedText
    defaultMessage="Loading Members..."
    id="mobile.loading_members"
    style={
      Object {
        "color": "rgba(61,60,64,0.6)",
      }
    }
  />
</Component>
`;

exports[`CustomList should match snapshot, renderSectionHeader 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
>
  <Component
    style={
      Object {
        "backgroundColor": "rgba(61,60,64,0.07)",
        "paddingLeft": 10,
        "paddingVertical": 2,
      }
    }
  >
    <Component
      style={
        Object {
          "color": "#3d3c40",
          "fontWeight": "600",
        }
      }
    >
      section_id
    </Component>
  </Component>
</Component>
`;

exports[`CustomList should match snapshot, renderSeparator 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "rgba(61,60,64,0.1)",
      "flex": 1,
      "height": 1,
    }
  }
/>
`;
